Framework/Core/System/IO/StreamReader.cs

```diff
@@ -87,7 +87,7 @@ public override int Peek()
             int nextChar;
 
             // If buffer need refresh take into account max UTF8 bytes if the next character is UTF8 encoded
-            if ((m_curBufPos                       == (m_curBufLen-1)) || 
+            if ((m_curBufPos                       >= (m_curBufLen-1)) || 
                ((m_buffer[m_curBufPos + 1] & 0x80) !=  0 && 
                 (m_curBufPos + 3                   >=  m_curBufLen)))

```

Comment:
Sounds like something in need of some comments explaining the situation so no-one is tempted to change things around